### PR TITLE
Uniform and optional use of expr()

### PR DESCRIPTION
**Pull request by hugoetchegoyen**

### CODE REVIEW

File: include/matchit.h [hunk 1]
```diff
@@ -104,13 +104,28 @@ namespace matchit
...
         constexpr auto expr(T const &v)
         {
             return nullary([&]
                            { return v; });
         }
 
+        template <typename T>
+        constexpr auto toNullary(T &&v)
+        {
+            if constexpr ( std::is_invocable_v<std::decay_t<T>> )
```
Comment (BowenFu) [inline include/matchit.h:123]: Same here. Run single-header.sh should fix this as well.

File: test/matchit/optexpr.cpp [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+constexpr auto eval1(std::tuple<char, T1, T2> const& exp)
+{
+    using namespace matchit;
+    Id<T1> i;
+    Id<T2> j;
+    return match(exp)   // no expr()
+    (
+        pattern | ds('+', i, j) | when ( (i + j) > 0 ) = i + j,
+        pattern | ds('-', i, j) | when ( true ) = i - j,
+        pattern | ds('*', i, j) | when ( i ) = i,
```
Comment (BowenFu) [inline test/matchit/optexpr.cpp:17]:
```suggestion
        pattern | ds('+', i, j) | when((i + j) > 0) = i + j,
        pattern | ds('-', i, j) | when(true) = i - j,
        pattern | ds('*', i, j) | when(i) = i,
```

File: test/matchit/optexpr.cpp [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+constexpr auto eval2(std::tuple<char, T1, T2> const& exp)
+{
+    using namespace matchit;
+    Id<T1> i;
+    Id<T2> j;
+    return match(exp)   // redundant expr()
+    (
+        pattern | ds('+', i, j) | when ( expr((i + j) > 0) ) = expr(i + j),
+        pattern | ds('-', i, j) | when ( expr(true) ) = expr(i - j),
+        pattern | ds('*', i, j) | when ( expr(i) ) = expr(i),
```
Comment (BowenFu) [inline test/matchit/optexpr.cpp:33]:
```suggestion
        pattern | ds('+', i, j) | when (expr((i + j) > 0)) = expr(i + j),
        pattern | ds('-', i, j) | when (expr(true)) = expr(i - j),
        pattern | ds('*', i, j) | when (expr(i)) = expr(i),
```

File: include/matchit.h (hunk 1 above)
Comment (hugoetchegoyen) [reply include/matchit.h]: done

File: develop/matchit/expression.h [hunk 4]
```diff
@@ -29,13 +29,28 @@ namespace matchit
                            { return *id; });
         }
 
+        template <typename T>
+        constexpr auto expr(Nullary<T> const &v)
+        {
+            return v;
```
[the remark this answers is not in the document]
Comment (hugoetchegoyen) [reply develop/matchit/expression.h]: I removed expr() for Nullarys and passed that funcionality to toNullary. Use of expr() es now exactly like it was before (for Ids and others) but is optional. This removes the possibility of expr(Id + Id) and expr(expr(...)). If you want to remove expr() from public use the only thing necessary is removing the line "using impl::expr" in namespace matchit.

**Comment by hugoetchegoyen** [comment]
Bowen, I don't know if you got my last comment. In the last commit I removed expr(() for Nullarys and made it toNullary(). So now expr() is the same as your original version (only for Ids and others) but it is optional. In order to remove it from public use it is only necessary to remove "using impl::expr" from namespace matchit.

File: develop/matchit/expression.h [hunk 5]
```diff
@@ -36,6 +36,25 @@ namespace matchit
                            { return v; });
         }
 
+        template <typename T> 
+        constexpr auto toNullary(Nullary<T> const &v)
```
Comment (BowenFu) [inline develop/matchit/expression.h:40]: Actually, this one is no longer needed, since it will be covered by the general version.

**Comment by BowenFu** [comment]
LGTM. I do not have other comments.

File: develop/matchit/expression.h (hunk 5 above)
Comment (hugoetchegoyen) [reply develop/matchit/expression.h]: You are right, I removed it.

**Pull request merged**